composer/datasets/streaming/dataset.py

```diff
@@ -237,91 +248,145 @@ def _insert_shard_samples(self, shard: int, part_min_id: int, part_max_id: int)
-        world = get_world()
-        part_shards, part_shards_to_download, part_min_id, part_max_id = self.index.get_partition(
-            world, self.batch_size)
+    def _download_shard(self, shard: int, shards_to_download: List[int]) -> int:
+        """Download the given shard.
 
-        if self.shuffle:
-            # Always process first shard first because other workers may be waiting on it
-            part_shards = np.array(part_shards)
-            np.random.shuffle(part_shards[1:])
-
-        for shard in part_shards:
-            # If this worker is in charge of downloading the shard, download it.
-            # Otherwise, wait until shard gets downloaded by another worker on this node
-            # This produces deterministic sample order.
-            basename = get_shard_basename(shard, compression_name=self.compression_scheme)
-            try:
-                self._download_file(basename, wait=(shard not in part_shards_to_download))
-            except Exception as e:
+        Args:
+            shard (int): Shard ID.
+            shards_to_download (List[int]): Which shards we are in charge of downloading.
+
+        Returns:
+            int: Shard ID.
+        """
+        basename = get_shard_basename(shard, self.compression_scheme)
+        wait = shard not in shards_to_download
+        self._download_file(basename, wait)
+        return shard
+
+    def _download_shards_pool(self, missing_shards: List[int], shards_to_download: List[int], min_id: int,
+                              max_id: int, num_processes: Optional[int]) -> None:
+        """Download and load the given missing shards.
+
+        Args:
+            missing_shards (List[int]): The missing shards.
+            shards_to_download (List[int]): List of shards to download by this worker.
+            min_id (int): Lowest sample ID of this partition.
+            max_id (int): Highest sample ID of this partition.
+            num_processes (Optional[int]): Number of concurrent shard downloads (ie, size of process pool). If None,
+                uses number of CPUs.
+        """
+        pool = Pool(num_processes)
+        download_shard = lambda shard: self._download_shard(shard, shards_to_download)
+        try:
+            for shard in pool.imap_unordered(download_shard, missing_shards):
+                self._load_shards([shard], min_id, max_id)
+        except Exception as e:
+            with self._lock:
                 self._download_status = _DownloadStatus.FAILED
                 self._download_exception = e
-            self._insert_shard_samples(shard, part_min_id, part_max_id)
-
+            return
         with self._lock:
             self._download_status = _DownloadStatus.DONE
 
-    def __len__(self) -> int:
-        """Get the length of the dataset.
+    def _download_shards_sequential(self, missing_shards: List[int], shards_to_download: List[int], min_id: int,
```

Comment:
Maybe we could split blocking and non-blocking download() into two separate methods, but all the setup work is hairy/in common

Reply:
Hmm, the code is 99% the same. Can you refactor the common parts into a shared helper? (This will be a moot point if we do only non-blocking downloads via a private method)

Reply:
Done